controllers/destinationProxy.js

```diff
@@ -0,0 +1,110 @@
+const networkHandlerFactory = require('../adapters/networkHandlerFactory');
+const { getPayloadData } = require("../adapters/network");
+const { TRANSFORMER_METRIC } = require("../v0/util/constant");
+const {
+  generateErrorObject,
+} = require("../v0/util");
+const stats = require("../util/stats");
+const match = require("match-json");
+const jsonDiff = require("json-diff");
+const logger = require("../logger");
+
+class DestProxyController {
+  /**
+   * Handler for testing the destination proxy
+   * @param {*} destination Destination name
+   * @param {*} ctx 
+   * @returns 
+   */
+  static handleProxyTestRequest = async (destination, ctx) => {
+    const { deliveryPayload: routerDeliveryPayload, destinationRequestPayload: routerDestReqPayload } = ctx.request.body;
+    let response;
+    try {
+      const destNetworkHandler = networkHandlerFactory.getNetworkHandler(
+        destination
+      );
+      
+      const proxyDestReqPayload = destNetworkHandler.prepareProxy(routerDeliveryPayload);
+      response = {
+        destinationRequestPayload: proxyDestReqPayload,
+      };
+
+      // Special handling required as Go and JavaScript encodes
+      // URL parameters differently
+      const { payloadFormat } = getPayloadData(routerDeliveryPayload.body);
+      if (payloadFormat === "FORM") {
+        // This is to make sure we encode `~` in the data coming from the router
+        // The result coming from the router is already a query parameter string
+        const routerBodyVal = new URLSearchParams(routerDestReqPayload.body);
+        routerDestReqPayload.body = routerBodyVal;
+  
+        const proxyBodyVal = new URLSearchParams();
+        proxyDestReqPayload.body.forEach(function(value, key) {
+          const encodeAsterisk = (x) => x.replace(/\*/g, "%2A");
+          // Router encodes `*` as well
+          proxyBodyVal.append(encodeAsterisk(key), encodeAsterisk(value));
```

Comment:
Any reason why we're encoding the key ? Because we don't see a similar thing being done while parameters are mapped in server-side

- https://github.com/rudderlabs/rudder-server/blob/1728d3ec011f2ea781a3305f29aba1e1ddf3eb67/router/network.go#L128-L132
- https://github.com/rudderlabs/rudder-server/blob/1728d3ec011f2ea781a3305f29aba1e1ddf3eb67/router/network.go#L150-L159

Reply:
The router's output string is converted into search params object for aligning the `~` encoding automatically. Comparing at the object level gives us an advantage when the order of the entries for some reason is not the same.